File: nim/ads/stack_linked.nim
type Node = ref object
  data: int
  head: Node
  next: Node

proc newStack(): Node =
  Node(data: -1)

proc newNode(data: int): Node =
  Node(data: data)

proc isEmpty(node: Node): bool =
  return node.head.isNil

proc push(stack: var Node, data: int) =
  let node = newNode(data)
  node.next = stack.head
  stack.head = node
  echo "Push: " & $data

proc pop(stack: var Node) =
  if (stack.head == nil):
    echo "Trying to pop from empty"
    return

  let temp = stack.head
  stack.head = temp.next
  echo "Pop: " & $temp.data

proc peek(stack: Node) =
  if (stack.isEmpty):
    echo "Stack Overflow"
    return

  let data = stack.head.data
  echo "Peek: " & $data

proc toStr(stack: Node) =
  var tmp = stack.head
  while (not tmp.isNil):
    stdout.write $tmp.data & " -> "
    tmp = tmp.next

proc main() =
  var stack = newStack()
  for i in 1..5: stack.push(i)

  echo()
  for _ in 0..2: stack.pop()

  echo()
  for _ in 0..1: stack.peek()

  echo()
  stack.toStr()
  echo()

when isMainModule:
  main()

File: clang/ads/stack_linked.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  int data;
  struct Node* next;
} Node;

typedef struct {
  Node* head;
} Stack;

int   peek(Stack* stack);
int   is_empty(Stack* stack);
void  pop(Stack* stack);
void  init_stack(Stack* stack);
void  push(Stack* stack, int new_data);
Node* create_node(int new_data);

Node* create_node(int new_data){
  Node* node = malloc(sizeof(Node));
  node->data = new_data;
  node->next = NULL;
  return node;
}

void init_stack(Stack* stack){
  stack->head = NULL;
}

int is_empty(Stack* stack){
  return stack->head == NULL;
}

void push(Stack* stack, int new_data){
  Node* node = create_node(new_data);
  if (!node){
    printf("\nStack Overflow");
    return;
  }

  node->next = stack->head;
  stack->head = node;
}

void pop(Stack* stack){
  if (is_empty(stack)){
    printf("\nStack Overflow");
    return;
  } else {
    Node* temp = stack->head;
    stack->head = temp->next;
    free(temp);
  }
}

int peek(Stack* stack){
  if (!is_empty(stack)){
    return stack->head->data;
  } else {
    printf("\nStack Overflow");
    return INT_MIN;
  }
}
